src/main/java/org/spongepowered/api/resource/Resource.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.resource;
+
+import org.spongepowered.api.data.persistence.DataFormat;
+import org.spongepowered.api.data.persistence.DataFormats;
+import org.spongepowered.api.data.persistence.DataView;
+import org.spongepowered.api.resource.metadata.MetadataSection;
+import org.spongepowered.api.resource.pack.Pack;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * A resource can represent any kind of loaded data. It can be a file on the
+ * filesystem, a network location, or even generated at runtime. Use
+ * {@link #inputStream()} to load the data held by a resource.
+ */
+public interface Resource extends AutoCloseable {
+
+    /**
+     * @return The {@link Pack pack}
+     */
+    Pack pack();
+
+    /**
+     * @return The {@link ResourcePath path}
+     */
+    ResourcePath path();
+
+    /**
+     * Returns the {@link InputStream} of this resource. Multiple calls to this
+     * method will not return a new object. To get a new object, get a new
+     * resource.
+     *
+     * @return The input stream
+     */
+    InputStream inputStream();
```

Comment:
I think this is going to cause some confusion.

Basically, what you're saying here is that this resource is backed by the input stream here - that's okay. What's a little more complicated here is that, at least as I read it, other methods are going to affect this stream, such as `newReader`, `readString` etc. and this may well trip people up.

I'm not sure as of the best way to convey this. It may be better to split this interface into two interfaces, perhaps `Managed` (with the `read*` methods that can rewind the stream if necessary) and `Unmanaged` (this method), and perhaps drop the `newReader` method altogether, if people want that they can get that from an "unmanaged" stream by creating the reader by themselves. 

Even if the impl implements both interfaces, allowing for casting by advanced users, I feel like making this separation may benefit everyone, because you're separating the idea of using a stream from the idea of using managed methods.

Alternatively, we just go all out on documentation. That may be preferred.

Reply:
So you'll find in the next commit I tossed most of the methods.

They are heavy handed in the API and anyone can easily do it on their own (wrap the stream in a reader/etc). The bigger problem, I feel, is Vanilla.

Vanilla caches the input stream on `SimpleResource` which is the only implementation. That said...SimpleResource is created by demand from the `ResourceManager` when queried. The API is strange because `PackResources` in Vanilla does not give you `Resource` objects (it really ought to) and instead just hands you an `InputStream`

I feel like we're close to beating this API framework in place, just some final bits and bobs to iron out.

Reply:
It honestly feels daft to me that the `InputStream` is tied to the `Resource` in the way that it is... but that's the hand we're dealt.

I'm guessing closing the resource closes the stream. It would probably be good to note that on the class docs.

Reply:
@dualspiral I've considered tossing Resource out completely.

Due to the nature of the re-creation and the fact that using it in a try-with-resources quite literally nukes the actual worthwhile contents (namely the stream), I am finding less and less usage of the actual object.

Reply:
Hmm or maybe not...bah this is difficult.

The issue is...`SimpleResource` doesn't open the stream up upon creation. While implementation shouldn't dictate API, we're dealing with nasty IO issues.

Thoughts?

Reply:
I was thinking about that. `Resource` contains callbacks to packs and paths... but I'm guessing those were used to get the resource in the first place?